ledger/simulation/simulator.go

```diff
@@ -0,0 +1,186 @@
+	SimulatorError        // the original error for internal use
+	External       string // the external error for public use
+}
+
+// ==============================
+// > Simulator
+// ==============================
+
+// Simulator is a transaction group simulator for the block evaluator.
+type Simulator struct {
+	ledger simulatorLedger
+}
+
+// MakeSimulator creates a new simulator from a ledger.
+func MakeSimulator(ledger *data.Ledger) *Simulator {
+	return &Simulator{
+		ledger: simulatorLedger{ledger, ledger.Latest()},
+	}
+}
+
+// check verifies that the transaction is well-formed and has valid or missing signatures.
+// An invalid transaction group error is returned if the transaction is not well-formed or there are invalid signatures.
+// To make things easier, we support submitting unsigned transactions and will respond whether signatures are missing.
+func (s Simulator) check(hdr bookkeeping.BlockHeader, txgroup []transactions.SignedTxn) (isMissingSigs bool, err error) {
+	specialAddresses := transactions.SpecialAddresses{
+		FeeSink:     hdr.FeeSink,
+		RewardsPool: hdr.RewardsPool,
+	}
+	consensusParams := config.Consensus[hdr.CurrentProtocol]
+
+	// Filter signed and unsigned transactions
+	var signedTxns []transactions.SignedTxn
+	for _, tx := range txgroup {
+		// If unsigned, check that they are well-formed
+		if verify.TxnIsMissingSig(&tx) {
+			isMissingSigs = true
+			if err = tx.Txn.WellFormed(specialAddresses, consensusParams); err != nil {
+				err = InvalidTxGroupError{SimulatorError{err}}
+				return
+			}
+			continue
+		}
+
+		// Otherwise add to the list of signed transactions
+		signedTxns = append(signedTxns, tx)
+	}
+
+	// Guard against no signed transactions
+	if len(signedTxns) == 0 {
+		return
+	}
+
+	// Verify the signed transactions are well-formed and have valid signatures
+	_, err = verify.TxnGroup(signedTxns, hdr, nil)
+	if err != nil {
+		err = InvalidTxGroupError{SimulatorError{err}}
+		return
+	}
+
+	return
+}
+
+func (s Simulator) evaluate(hdr bookkeeping.BlockHeader, stxns []transactions.SignedTxn) (*ledgercore.ValidatedBlock, error) {
+	// s.ledger has 'StartEvaluator' because *data.Ledger is embedded in the simulatorLedger
+	// and data.Ledger embeds *ledger.Ledger
+	eval, err := s.ledger.StartEvaluator(hdr, len(stxns), 0)
+	if err != nil {
+		return nil, err
+	}
+
+	group := transactions.WrapSignedTxnsWithAD(stxns)
+
+	err = eval.TransactionGroup(group)
+	if err != nil {
+		return nil, EvalFailureError{SimulatorError{err}}
+	}
+
+	// Finally, process any pending end-of-block state changes.
+	vb, err := eval.GenerateBlock()
+	if err != nil {
+		return nil, err
+	}
+
+	return vb, nil
+}
+
+// Simulate simulates a transaction group using the simulator. Will error if the transaction group is not well-formed.
+func (s Simulator) Simulate(txgroup []transactions.SignedTxn) (vb *ledgercore.ValidatedBlock, missingSignatures bool, err error) {
+	prevBlockHdr, err := s.ledger.BlockHdr(s.ledger.start)
```

Comment:
@jannotti do you think this `Simulate` method should accept a `BlockHeader` to be consistent with `Ledger.StartEvaluator`?